File: node/src/langchain.ts
import { type StructuredTool, tool as langchainTool } from 'langchain'

import { ListToolkit } from './toolkit.js'
import { type Tool } from './tools.js'
import { errorMessage } from './util.js'

export class AgentMailToolkit extends ListToolkit<StructuredTool> {
    protected buildTool(tool: Tool): StructuredTool {
        return langchainTool(
            async (args: Record<string, any>) => {
                try {
                    return JSON.stringify(await tool.func(this.client, args), null, 2)
                } catch (err) {
                    // The installed @langchain/core (1.1.x) has no `ToolException` class -
                    // that's a langchain-python construct, not JS. The framework-native
                    // mechanism here is to let the tool function's promise reject:
                    // DynamicStructuredTool.call() propagates the rejection rather than
                    // swallowing it, so the calling agent/executor sees a real failure
                    // instead of an error string dressed up as a successful result.
                    throw new Error(errorMessage(err))
                }
            },
            {
                name: tool.name,
                description: tool.description,
                schema: tool.paramsSchema,
            }
        )
    }
}

File: python/src/agentmail_toolkit/langchain.py
from typing import Optional
from langchain.tools import BaseTool, ToolException, tool as langchain_tool
from agentmail import AgentMail

from .toolkit import Toolkit
from .tools import Tool
from .util import api_error_message


class AgentMailToolkit(Toolkit[BaseTool]):
    def __init__(self, client: Optional[AgentMail] = None):
        super().__init__(client)

    def _build_tool(self, tool: Tool):
        def runnable(**kwargs):
            try:
                return tool.func(self.client, kwargs)
            except Exception as e:
                raise ToolException(api_error_message(e)) from e

        return langchain_tool(
            name_or_callable=tool.name,
            description=tool.description,
            args_schema=tool.params_schema,
            runnable=runnable,
            handle_tool_error=True,
        )
